tests/core/test_graph.py

```diff
@@ -1,54 +1,367 @@
+def test_nodes(simple_graph):
+    i = 0
+    for node in simple_graph.nodes:
+        assert node in ["a", "b", "c"]
+        i += 1
+    assert i == 3
+
+
+def test_edges(simple_graph):
+    i = 0
+    for edge in simple_graph.edges:
+        assert edge in [("a", "b"), ("b", "c"), ("a", "c")]
+        i += 1
+    assert i == 3
 
 
 def test_successors(simple_graph):
     s = [v for v in simple_graph.successors("a")]
     assert s == ["b", "c"]
 
+    s = [v for v in simple_graph.successors("c")]
+    assert len(s) == 0
+
 
 def test_predecessors(simple_graph):
     s = [v for v in simple_graph.predecessors("b")]
     assert s == ["a"]
 
+    s = [v for v in simple_graph.predecessors("a")]
+    assert len(s) == 0
+
 
 def test_is_edge(simple_graph):
     assert simple_graph.is_edge("a", "b")
+    assert not simple_graph.is_edge("b", "a")
+    assert simple_graph.is_edge("a", "c")
+    assert not simple_graph.is_edge("c", "a")
+    assert simple_graph.is_edge("b", "c")
+    assert not simple_graph.is_edge("c", "b")
+
+
+def test_get_sparse_adj_matrix(simple_graph):
+    adj = simple_graph.get_sparse_adj_matrix()
+    assert adj.shape == (3, 3)
+    assert adj.nnz == 3
+
+    edge_weight = torch.tensor([[1], [1], [2]])
+    simple_graph.data["edge_weight"] = edge_weight
+    weighted_adj = simple_graph.get_sparse_adj_matrix("edge_weight")
+    assert weighted_adj.shape == (3, 3)
+    assert weighted_adj.nnz == 3
+    assert isinstance(weighted_adj, s.coo_matrix)
+    assert weighted_adj.data[0] == 1
+    assert weighted_adj.data[1] == 1
+    assert weighted_adj.data[2] == 2
+
+
+def test_degrees(simple_graph):
+    in_degrees = simple_graph.degrees("in")
+    assert in_degrees["a"] == 0
+    assert in_degrees["b"] == 1
+    assert in_degrees["c"] == 2
+
+    out_degrees = simple_graph.degrees("out")
+    assert out_degrees["a"] == 2
+    assert out_degrees["b"] == 1
+    assert out_degrees["c"] == 0
+
+
+def test_in_degrees(simple_graph):
+    in_degrees = simple_graph.in_degrees
+    assert in_degrees["a"] == 0
+    assert in_degrees["b"] == 1
+    assert in_degrees["c"] == 2
+
+
+def test_out_degrees(simple_graph):
+    out_degrees = simple_graph.out_degrees
+    assert out_degrees["a"] == 2
+    assert out_degrees["b"] == 1
+    assert out_degrees["c"] == 0
+
+
+def test_get_laplacian(simple_graph):
+    laplacian = simple_graph.get_laplacian()
+    assert laplacian.shape == (3, 3)
+    assert laplacian.nnz == 6
+    assert isinstance(laplacian, s.coo_matrix)
+    assert laplacian.data[0] == -1
+    assert laplacian.data[1] == -1
+    assert laplacian.data[2] == -1
+    assert laplacian.data[3] == 2
+    assert laplacian.data[4] == 1
+    assert laplacian.data[5] == 0
+
+
+def test_add_node_ohe(simple_graph):
+    simple_graph.add_node_ohe("node_ohe")
+    assert simple_graph.data["node_ohe"].shape == (3, 3)
+    assert torch.equal(simple_graph.data["node_ohe"], torch.eye(3))
+
+    simple_graph.add_node_ohe("node_ohe_dim_small", 2)
```

Comment:
What is the benefit of specifying `dim` here? Is the below behaviour something that we want?

Reply:
The idea was that we can specify the dimension of the one-hot encoding. However, we can probably remove this function, as the user can now directly access attributes in the underlying `Data` frame, which allows to add one hot encodings as well.